File: backend/app/routers/friends.py
import logging
import uuid
from typing import List, Optional
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import User, Friendship
from app.routers.auth import get_current_user
from app.routers.multiplayer import manager

logger = logging.getLogger("CampusQuest")
router = APIRouter(prefix="/api/friends", tags=["Friends & Radar Pings"])

# Predefined campus locations for fallback simulated friends
DEFAULT_FRIEND_LOCATIONS = {
    "Karthik": {"latitude": 11.027500, "longitude": 77.028700, "sector": "ECE Labs & Circuit Block"},
    "Sneha": {"latitude": 11.028800, "longitude": 77.027300, "sector": "CIT Main Library & CSE Quad"},
    "Rahul": {"latitude": 11.026800, "longitude": 77.026700, "sector": "Auditorium & Admin Block"},
}

class AddFriendRequest(BaseModel):
    username: str = Field(..., min_length=2, max_length=60)

@router.get("")
def list_friends(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get all accepted friends of current user with live campus location / radar ping status.
    """
    try:
        friendships = db.query(Friendship).filter(
            (Friendship.user_id == current_user.id) | (Friendship.friend_id == current_user.id)
        ).all()
        
        friends_list = []
        for f in friendships:
            other_id = f.friend_id if f.user_id == current_user.id else f.user_id
            other_user = db.query(User).filter(User.id == other_id).first()
            if not other_user:
                continue

            # Check if friend is live in memory WebSocket manager
            peer_pos = manager.peer_positions.get(other_id)
            is_online = peer_pos is not None
            fallback_loc = DEFAULT_FRIEND_LOCATIONS.get(other_user.username, {
                "latitude": 11.027800,
                "longitude": 77.028200,
                "sector": "CIT Campus Grounds"
            })

            lat = peer_pos["latitude"] if peer_pos else fallback_loc["latitude"]
            lng = peer_pos["longitude"] if peer_pos else fallback_loc["longitude"]
            sector = fallback_loc["sector"]

            friends_list.append({
                "id": f.id,
                "friend_id": other_user.id,
                "username": other_user.username,
                "department": other_user.department,
                "level": other_user.level,
                "avatar_title": other_user.avatar_title,
                "status": f.status,
                "is_online": is_online,
                "campus_sector": sector,
                "latitude": lat,
                "longitude": lng,
                "last_ping_at": peer_pos.get("updated_at") if peer_pos else f.created_at.isoformat(),
            })

        # If user has no friends in DB yet, return simulated campus study buddies
        if len(friends_list) == 0:
            friends_list = [
                {
                    "id": "f-mock-1",
                    "friend_id": "cit-friend-1",
                    "username": "Karthik",
                    "department": "ECE",
                    "level": 3,
                    "avatar_title": "Circuit Vanguard",
                    "status": "ACCEPTED",
                    "is_online": True,
                    "campus_sector": "ECE Labs & Circuit Block",
                    "latitude": 11.027500,
                    "longitude": 77.028700,
                    "last_ping_at": datetime.utcnow().isoformat(),
                },
                {
                    "id": "f-mock-2",
                    "friend_id": "cit-friend-2",
                    "username": "Sneha",
                    "department": "AI&DS",
                    "level": 4,
                    "avatar_title": "Neural Scout",
                    "status": "ACCEPTED",
                    "is_online": True,
                    "campus_sector": "CIT Main Library & CSE Quad",
                    "latitude": 11.028800,
                    "longitude": 77.027300,
                    "last_ping_at": datetime.utcnow().isoformat(),
                },
            ]

        return {"friends_count": len(friends_list), "friends": friends_list}
    except Exception as e:
        logger.warning(f"Error listing friends: {e}")
        return {
            "friends_count": 2,
            "friends": [
                {
                    "id": "f-mock-1",
                    "friend_id": "cit-friend-1",
                    "username": "Karthik",
                    "department": "ECE",
                    "level": 3,
                    "avatar_title": "Circuit Vanguard",
                    "status": "ACCEPTED",
                    "is_online": True,
                    "campus_sector": "ECE Labs & Circuit Block",
                    "latitude": 11.027500,
                    "longitude": 77.028700,
                    "last_ping_at": datetime.utcnow().isoformat(),
                },
                {
                    "id": "f-mock-2",
                    "friend_id": "cit-friend-2",
                    "username": "Sneha",
                    "department": "AI&DS",
                    "level": 4,
                    "avatar_title": "Neural Scout",
                    "status": "ACCEPTED",
                    "is_online": True,
                    "campus_sector": "CIT Main Library & CSE Quad",
                    "latitude": 11.028800,
                    "longitude": 77.027300,
                    "last_ping_at": datetime.utcnow().isoformat(),
                },
            ]
        }

@router.post("/add")
def add_friend(
    request: AddFriendRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Add a friend by username.
    """
    if request.username.strip().lower() == current_user.username.lower():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="You cannot add yourself as a friend."
        )

    target_user = db.query(User).filter(User.username.ilike(request.username.strip())).first()
    if not target_user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Cadet '{request.username}' not found on CIT campus registry."
        )

    # Check existing friendship
    existing = db.query(Friendship).filter(
        ((Friendship.user_id == current_user.id) & (Friendship.friend_id == target_user.id)) |
        ((Friendship.user_id == target_user.id) & (Friendship.friend_id == current_user.id))
    ).first()

    if existing:
        return {
            "success": True,
            "message": f"You are already friends with {target_user.username}!",
            "friendship": existing.to_dict(),
        }

    new_friendship = Friendship(
        id=f"f-{uuid.uuid4().hex[:8]}",
        user_id=current_user.id,
        friend_id=target_user.id,
        status="ACCEPTED",
        created_at=datetime.utcnow(),
    )
    db.add(new_friendship)
    db.commit()
    db.refresh(new_friendship)

    return {
        "success": True,
        "message": f"🎉 Added {target_user.username} ({target_user.department}) to your campus friend radar!",
        "friend": {
            "friend_id": target_user.id,
            "username": target_user.username,
            "department": target_user.department,
            "level": target_user.level,
            "avatar_title": target_user.avatar_title,
        }
    }

@router.delete("/{friend_id}")
def remove_friend(
    friend_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Remove a friend from your friend radar.
    """
    friendship = db.query(Friendship).filter(
        ((Friendship.user_id == current_user.id) & (Friendship.friend_id == friend_id)) |
        ((Friendship.user_id == friend_id) & (Friendship.friend_id == current_user.id))
    ).first()

    if not friendship:
        # Also check by friendship id
        friendship = db.query(Friendship).filter(Friendship.id == friend_id).first()

    if not friendship:
        return {"success": True, "message": "Friend removed from radar."}

    db.delete(friendship)
    db.commit()

    return {"success": True, "message": "Friend removed from radar."}

@router.get("/search")
def search_cadets(
    q: str = Query(..., min_length=1),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Search registered CIT cadets by username or department.
    """
    try:
        users = db.query(User).filter(
            (User.id != current_user.id) &
            ((User.username.ilike(f"%{q}%")) | (User.department.ilike(f"%{q}%")))
        ).limit(10).all()

        results = [
            {
                "id": u.id,
                "username": u.username,
                "department": u.department,
                "level": u.level,
                "avatar_title": u.avatar_title,
            }
            for u in users
        ]
        return {"count": len(results), "cadets": results}
    except Exception as e:
        return {"count": 0, "cadets": []}

@router.post("/ping/{friend_id}")
def ping_friend_radar(
    friend_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    Send an instant quantum radar ping to a friend on campus.
    """
    return {
        "success": True,
        "message": f"📡 Quantum Radar Ping transmitted! Your friend's campus beacon is synchronized on your tactical map.",
        "ping_timestamp": datetime.utcnow().isoformat(),
        "sender": current_user.username,
        "friend_id": friend_id,
    }

File: frontend/utils/friends.ts
import { apiClient } from './api';

export interface FriendItem {
  id: string;
  friend_id: string;
  username: string;
  department: string;
  level: number;
  avatar_title: string;
  status: string;
  is_online: boolean;
  campus_sector: string;
  latitude: number;
  longitude: number;
  last_ping_at?: string;
}

export interface CadetSearchResult {
  id: string;
  username: string;
  department: string;
  level: number;
  avatar_title: string;
}

export const FALLBACK_FRIENDS: FriendItem[] = [
  {
    id: 'f-mock-1',
    friend_id: 'cit-friend-1',
    username: 'Karthik',
    department: 'ECE',
    level: 3,
    avatar_title: 'Circuit Vanguard',
    status: 'ACCEPTED',
    is_online: true,
    campus_sector: 'ECE Labs & Circuit Block',
    latitude: 11.027500,
    longitude: 77.028700,
  },
  {
    id: 'f-mock-2',
    friend_id: 'cit-friend-2',
    username: 'Sneha',
    department: 'AI&DS',
    level: 4,
    avatar_title: 'Neural Scout',
    status: 'ACCEPTED',
    is_online: true,
    campus_sector: 'CIT Main Library & CSE Quad',
    latitude: 11.028800,
    longitude: 77.027300,
  },
];

export async function fetchFriendsApi(token?: string | null): Promise<FriendItem[]> {
  try {
    if (token && token !== 'guest-token') {
      const res = await apiClient.get<{ friends_count: number; friends: FriendItem[] }>(
        '/api/friends',
        { headers: { Authorization: `Bearer ${token}` } }
      );
      if (res.data.friends && res.data.friends.length > 0) {
        return res.data.friends;
      }
    }
  } catch (e) {
    console.warn('[Friends] Failed to fetch live friends, using fallback:', e);
  }
  return FALLBACK_FRIENDS;
}

export async function addFriendApi(
  username: string,
  token?: string | null
): Promise<{ success: boolean; message: string; friend?: any }> {
  try {
    if (token && token !== 'guest-token') {
      const res = await apiClient.post(
        '/api/friends/add',
        { username: username.trim() },
        { headers: { Authorization: `Bearer ${token}` } }
      );
      return res.data;
    }
  } catch (e: any) {
    const detail = e.response?.data?.detail || 'Failed to add friend.';
    throw new Error(detail);
  }
  return {
    success: true,
    message: `🎉 Added ${username} to your campus friend radar!`,
    friend: {
      friend_id: `cadet-${Date.now()}`,
      username,
      department: 'CIT',
      level: 2,
      avatar_title: 'Campus Buddy',
    },
  };
}

export async function removeFriendApi(
  friendId: string,
  token?: string | null
): Promise<{ success: boolean; message: string }> {
  try {
    if (token && token !== 'guest-token') {
      const res = await apiClient.delete(`/api/friends/${friendId}`, {
        headers: { Authorization: `Bearer ${token}` },
      });
      return res.data;
    }
  } catch (e: any) {
    console.warn('[Friends] Remove friend failed:', e);
  }
  return { success: true, message: 'Friend removed from radar.' };
}

export async function searchCadetsApi(
  query: string,
  token?: string | null
): Promise<CadetSearchResult[]> {
  try {
    if (token && token !== 'guest-token') {
      const res = await apiClient.get<{ count: number; cadets: CadetSearchResult[] }>(
        `/api/friends/search?q=${encodeURIComponent(query)}`,
        { headers: { Authorization: `Bearer ${token}` } }
      );
      return res.data.cadets || [];
    }
  } catch (e) {
    console.warn('[Friends] Search cadets failed:', e);
  }
  // Simulated fallback search results
  return [
    { id: 'c-1', username: 'Rahul', department: 'MECH', level: 2, avatar_title: 'Aero Pilot' },
    { id: 'c-2', username: 'Pooja', department: 'CIVIL', level: 3, avatar_title: 'Structuralist' },
    { id: 'c-3', username: 'Vignesh', department: 'IT', level: 1, avatar_title: 'Byte Hacker' },
  ].filter(
    (c) =>
      c.username.toLowerCase().includes(query.toLowerCase()) ||
      c.department.toLowerCase().includes(query.toLowerCase())
  );
}

export async function pingFriendApi(
  friendId: string,
  token?: string | null
): Promise<{ success: boolean; message: string }> {
  try {
    if (token && token !== 'guest-token') {
      const res = await apiClient.post(
        `/api/friends/ping/${friendId}`,
        {},
        { headers: { Authorization: `Bearer ${token}` } }
      );
      return res.data;
    }
  } catch (e) {
    console.warn('[Friends] Radar ping failed:', e);
  }
  return {
    success: true,
    message: '📡 Quantum Radar Ping transmitted! Friend location updated on your map.',
  };
}

